--- a/meson.build	2021-04-19 01:17:13.000000000 +1000
+++ b/meson.build	2021-04-19 04:03:13.141528070 +1000
@@ -26,7 +26,7 @@
 pipewire = dependency('libpipewire-0.3', version: '>= 0.3.2')
 wayland_client = dependency('wayland-client')
 wayland_protos = dependency('wayland-protocols', version: '>=1.14')
-iniparser = cc.find_library('iniparser', has_headers: ['iniparser.h', 'dictionary.h'])
+iniparser = cc.find_library('iniparser4', has_headers: ['iniparser.h', 'dictionary.h'])
 
 epoll = dependency('', required: false)
 if (not cc.has_function('timerfd_create', prefix: '#include <sys/timerfd.h>') or
