--- a/thunar/thunar-session-client.c
+++ b/thunar/thunar-session-client.c
@@ -42,7 +42,6 @@
 #include "thunar/thunar-private.h"
 #include "thunar/thunar-session-client.h"
 
-#include <gdk/gdkx.h>
 #include <glib/gstdio.h>
 
 
@@ -170,9 +169,6 @@ thunar_session_client_connect (ThunarSessionClient *session_client,
   if (G_UNLIKELY (session_client->connection == NULL))
     return FALSE;
 
-  /* tell GDK about our new session id */
-  gdk_x11_set_sm_client_id (id);
-
   /* remember the returned client id */
   session_client->id = id;
 
--- a/configure
+++ b/configure
@@ -21504,7 +21504,7 @@
 then :
   enableval=$enable_wallpaper_plugin; ac_bm_thunar_plugin_wallpaper=$enableval
 else $as_nop
-  ac_bm_thunar_plugin_wallpaper=yes
+  ac_bm_thunar_plugin_wallpaper=no
 fi

 { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking whether to build the thunar-wallpaper plugin" >&5
