diff --git a/nvidia-xinitrc b/nvidia-xinitrc
index d7f38ce..dfab77e 100644
--- a/nvidia-xinitrc
+++ b/nvidia-xinitrc
@@ -24,7 +24,7 @@ if [[ -f "$usermodmap" ]]; then
     xmodmap "$usermodmap"
 fi
 
-export LD_LIBRARY_PATH=/usr/lib64/nvidia/:/usr/lib32/nvidia:/usr/lib:${LD_LIBRARY_PATH}
+export LD_LIBRARY_PATH=/usr/lib64/opengl/nvidia/lib/:/usr/lib/opengl/nvidia/lib:/usr/lib:${LD_LIBRARY_PATH}
 
 # load additional configs
 if [[ -d /etc/X11/xinit/nvidia-xinitrc.d ]] ; then
diff --git a/nvidia-xorg.conf b/nvidia-xorg.conf
index 46b345b..3ece152 100644
--- a/nvidia-xorg.conf
+++ b/nvidia-xorg.conf
@@ -1,11 +1,6 @@
 Section "Files"
-  ModulePath "/usr/lib/nvidia"
-  ModulePath "/usr/lib32/nvidia"
-  ModulePath "/usr/lib32/nvidia/xorg/modules"
-  ModulePath "/usr/lib32/xorg/modules"
-  ModulePath "/usr/lib64/nvidia/xorg/modules"
-  ModulePath "/usr/lib64/nvidia/xorg"
+  ModulePath "/usr/lib64/extensions/nvidia"
   ModulePath "/usr/lib64/xorg/modules"
 EndSection
 
 Section "ServerLayout"

