--- brightray.gyp.orig	2016-08-29 08:43:35.554632534 +1000
+++ brightray.gyp	2016-08-29 08:43:40.824399610 +1000
@@ -1,7 +1,7 @@
 {
   'variables': {
     # The libraries brightray will be compiled to.
-    'linux_system_libraries': 'gtk+-2.0 dbus-1 x11 xi xcursor xdamage xrandr xcomposite xext xfixes xrender xtst xscrnsaver gmodule-2.0 nss'
+    'linux_system_libraries': 'gtk+-3.0 dbus-1 x11 xi xcursor xdamage xrandr xcomposite xext xfixes xrender xtst xscrnsaver gmodule-2.0 nss'
   },
   'includes': [
     'filenames.gypi',
