--- a/third_party/WebKit/Source/platform/platform_generated.gyp	2016-08-01 02:51:21.000131730 +1000
+++ b/third_party/WebKit/Source/platform/platform_generated.gyp	2016-08-01 03:01:08.996221199 +1000
@@ -174,6 +174,13 @@
       'dependencies': [
         '<(DEPTH)/third_party/icu/icu.gyp:icuuc#host',
       ],
+	  'conditions': [
+	    ['use_system_icu==1', {
+		  'defines': [
+		    'USING_SYSTEM_ICU',
+		  ],
+		}],
+	  ],
     },
   ],
 }
